File: LectureCodes/NumQuad/simpson/Matlab/simpson.m
function res = simpson(fnct,a,b,N)
% Numerical quadrature based on Simpson rule
% fnct handle to y = f(x)
% a,b bounds of integration interval
% N+1 = number of equidistant integration points (can be a vector)

res = [];
for n = N
  h = (b-a)/n;
  x = (a:h/2:b);
  fv = feval(fnct,x);
  val = sum(h*(fv(1:2:end-2)+4*fv(2:2:end-1)+fv(3:2:end)))/6;
  res = [res; h, val];
end

File: LectureCodes/NumQuad/simpson/Python/simpson.py
import numpy as np


def simpson(f, a, b, n):
    """ Numerical quadrature using Simpson rule """
    x = np.linspace(a, b, 2 * n + 1)
    h = (b - a) / n
    fv = f(x)
    return np.sum(h * (fv[:-2:2] + 4 * fv[1::2] + fv[2::2])) / 6


def main():
    def f(x):
        return np.sqrt(1 - x**2)

    print('pi / 2 - Integral of sqrt(1 - x^2) in domain [-1, 1] = {}'.format(
          np.pi / 2 - simpson(f, -1.0, 1.0, 1000)))


if __name__ == '__main__':
    main()
